[BE] feat: 중복 요청 방지 기능 구현

nayonsoso opened the pull request:
<!-- 제목: [BE/FE/All] (기능 등) -->
<!-- 아래에 이슈 번호를 매겨주세요 -->

- resolves #521 

---

### 🚀 어떤 기능을 구현했나요 ?
- 중복 요청 방지 기능을 구현했습니다.

### 🔥 어떻게 해결했나요 ?
- 커밋 따라서 쭉쭉 읽으시면 이해가 편하실거예요.
- 그리고 테스트 코드도 쭉 읽으면 완전 이해 완료하실 수 있습니다💡
- 요청을 받아서, 중복 로직 방지 로직을 수행하는 곳은 "인터셉터"입니다.
- 요청 거부 전략으로는 POST 요청에 대해서, 
- ‼️1초 동안 동일한 URI, public IP, User-Agent 의 조합으로 3번 초과 요청이 오면‼️
- 그 이후부터 429 응답을 내려줍니다.

![image](https://github.com/user-attachments/assets/fed8737b-ff9a-4bd4-8554-7fd3847a4913)
![image](https://github.com/user-attachments/assets/fe4fab2d-7696-44ca-8bea-31783a0a235d)

⬆️ 위 사진은 Jmeter (반복 요청을 보낼 수 있는 도구) 를 사용해서 1초동안 여러번의 요청을 보내본 결과입니다.

### 📝 어떤 부분에 집중해서 리뷰해야 할까요?
- 중복 요청 거부 로직에 대해서, 이해가 어려운 부분들 있다면 바로 질문해주세요

<br>

---

<br>

### 🔮 미래 계획

#### 1) dev-ec2 에 redis 환경 만들어주기
필요한 것은 "작동하는 redis" 와 "java application 과 redis 의 연결 설정" 인데요,
전자는 docker 를 사용해서 이미 갖춰두었고,
이 PR 머지하기 전에 application-dev.yml 에 아래 코드 추가할 예정입니다.
```yml
redis:
  host: localhost
  port: 6379
```

#### 2) prod-ec2-a / prod-ec2-b 에 redis 환경 만들어주기
redis 만을 위한 ec2 를 만들어준 다음, (역시나 도커로 편하게 redis 다운 &. 실행시킬 생각입니다)
application-prod.yml 에 아래 코드를 추가해줍니다.

```yml
redis:
  host: {그 ec2의 private ip}
  port: 6379
```
![image](https://github.com/user-attachments/assets/c1c3f3b5-1987-44a2-91ce-f83113859d84)

보안그룹은 이미 만들어져있는 project-cache 를 새로 만들 ec2 에 달아주면 될 것 같습니다

---

### 로컬에서 실행하는 방법

Redis 를 다운받고, 실행한다. (Redis 다운받는 법은 인터넷에 많음)
아래 명령어를 입력했을 때, 버전이 나오면 다운로드 된 것이다.

```bash
redis-server --version
```

이제 실행한다. 
```bash
brew services start redis
```

⚠️ 한가지 이슈 : RedisTemplate 을 빈으로 등록했기 때문에, 
redis 를 실행하지 않은 상태에서 local Application을 실행하면 오류가 뜨며 실행되지 않는다🙀

nayonsoso committed:
build: 의존성 추가

nayonsoso committed:
chore: 개발 환경 redis 설정

nayonsoso committed:
feat: 레디스 설정을 빈으로 등록

nayonsoso committed:
feat: 중복 요청 처리 인터셉터 생성

- URI, IP, UserAgent 를 기준으로 1초에 3개보다 많은 요청이 오는 경우, 429를 응답한다. 즉, 4번째 요청부터 거부된다.

nayonsoso committed:
feat: 중복 요청 처리 인터셉터 등록

nayonsoso committed:
feat: 예외 핸들러 추가

nayonsoso committed:
test: 중복 예외 처리 인터셉터 테스트 작성

nayonsoso committed:
refactor: value를 업데이트할 때 만료 시간을 다시 설정하도록 수정

nayonsoso committed:
style: 가독성을 위한 개행 수정

donghoony commented on backend/src/main/resources/application.yml:
```diff
@@ -34,3 +34,7 @@ cors:
+
+redis:
+  host: localhost
+  port: 6379
```
라인 끝 개행 부탁해요~

nayonsoso replied:
반완~

donghoony commented on backend/src/main/java/reviewme/config/RedisConfig.java:
```diff
@@ -0,0 +1,31 @@
+@Configuration
+public class RedisConfig {
+
+    @Value("${redis.host}")
+    private String host;
+
+    @Value("${redis.port}")
+    private int port;
```
`@Value` 대신 `@ConfigurationProperties`를 사용하면 어떨까요? yml에서 직접 읽어올 수도 있고, 응집도가 올라간다고 생각해요

nayonsoso replied:
반완!

donghoony commented on backend/src/main/java/reviewme/global/DuplicateRequestInterceptor.java:
```diff
@@ -0,0 +1,57 @@
+public class DuplicateRequestInterceptor implements HandlerInterceptor {
+
+    private static final int MAX_FREQUENCY = 3;
+    private static final Duration DURATION = Duration.of(1, ChronoUnit.SECONDS);
```
`Duration.ofSeconds(1)` 은 어떨까요 ?

donghoony replied:
지금 보니 1초를 `DURATION`이라는 변수에 저장했네요. `SECOND`로 의미를 담는 건 어떨까요 ?

nayonsoso replied:
반완완~

donghoony commented on backend/src/main/java/reviewme/global/DuplicateRequestInterceptor.java:
```diff
@@ -0,0 +1,55 @@
+            throw new TooManyDuplicateRequestException(key);
+        }
+        redisTemplate.opsForValue().set(key, frequency + 1, DURATION);
+        return true;
```
이부분은 한 번 생각해보겠습니다 🤔

donghoony commented on backend/src/main/java/reviewme/global/GlobalExceptionHandler.java:
```diff
@@ -50,6 +51,11 @@ public ProblemDetail handleDataConsistencyException(DataInconsistencyException e
         return ProblemDetail.forStatusAndDetail(HttpStatus.INTERNAL_SERVER_ERROR, ex.getErrorMessage());
     }
 
+    @ExceptionHandler(TooManyDuplicateRequestException.class)
```
(네이밍) `Duplicate`는 없어도 될 듯합니다! `TooManyRequestException`으로도 충분하다고 생각해요

nayonsoso replied:
`TooManyRequestException` 은 '요청 자체가 너무 많아서' 서버가 처리할 수 없다는 뉘앙스인 것 같아요,
이 예외가 발생하는 상황은 그런게 아니니까 구분하기 위해서라도 Duplicate를 붙이고 싶어요!

donghoony replied:
다른 단어를 선택하는 건 어떨까요? `duplicate`는 정말 완벽히 같은 복제본의 느낌인데, 본문이 다르더라도 많은 요청이 막힐 수 있는 부분이라서요. 🤔

donghoony replied:
`Many`와 `Duplicate`의 의미가 중복돼 클래스 이름이 어색해서 남긴 리뷰였습니다~

nayonsoso replied:
> 본문이 다르더라도 많은 요청이 막힐 수 있는 부분이라서요.
> Many와 Duplicate의 의미가 중복돼 클래스 이름이 어색해서 남긴 리뷰였습니다~

아하 그럴수도 있겠네요..! 이해됩니다 ㅎㅎ

`TooManyRequestException`로 하겠습니다~
찾아보니 429 자체에 "특정 사용자"라는 뉘앙스가 있기도 하네요 

![image](https://github.com/user-attachments/assets/76258aaa-402f-480a-8883-01770c62470d)

donghoony commented on backend/src/main/java/reviewme/global/exception/TooManyDuplicateRequestException.java:
```diff
@@ -0,0 +1,12 @@
+
+    public TooManyDuplicateRequestException(String requestKey) {
+        super("짧은 시간 안에 너무 많은 동일한 요청이 일어났어요");
+        log.info("Too many duplicate request arrived: {}", requestKey);
```
`duplicate` 없어도 될 듯해요 2

nayonsoso replied:
Ditto!

hyeonji1220 commented on backend/src/main/java/reviewme/global/exception/TooManyDuplicateRequestException.java:
```diff
@@ -0,0 +1,12 @@
+
+    public TooManyDuplicateRequestException(String requestKey) {
+        super("짧은 시간 안에 너무 많은 동일한 요청이 일어났어요");
+        log.info("Too many duplicate request arrived: {}", requestKey);
```
소소한 제안)
arrived 대신 클라이언트-서버 관점의 단어로 received가 어떨까요?

nayonsoso replied:
반완~

hyeonji1220 commented on backend/src/main/java/reviewme/global/DuplicateRequestInterceptor.java:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        String key = generateRequestKey(request);
+        Object value = redisTemplate.opsForValue().get(key);
+        if (value == null) {
+            redisTemplate.opsForValue().set(key, 1, DURATION);
+            return true;
+        }
+
+        if (!(value instanceof Integer)) {
+            throw new RequestFrequencyNonNumericException(value);
+        }
```
찾아보니 있으면 기존 value에 1 증가 시켜주는 일을 해주는 `redisTemplate.opsForValue().increment(key)` 라는 메서드가 있는 것 같은데 직접 구현한 이유가 있을까요?

donghoony replied:
이거 좋네요 👍🏻

nayonsoso replied:
이렇게 한 배경은...

redisTemplate.opsForValue().increment(key) 는 value 에 해당하는 값이 numeric 이 아니면 예외가 발생해요. 그런데 저는 RedisTemplate 의 value 는 확장성을 위해서 Integer 나 Long 이 아니라 Object 여야 한다고 생각해요!

였지만 커비의 제안으로 RedisTemplate<String, Long> 로 형식을 바꾸게 됨으로써 저 메서드를 사용하도록 바꿨습니다🫡

hyeonji1220 commented on backend/src/main/java/reviewme/global/DuplicateRequestInterceptor.java:
```diff
@@ -0,0 +1,55 @@
+    private final RedisTemplate<String, Object> redisTemplate;
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
```
value로는 무조건 숫자가 오는 것 같은데 int 나 long이 아닌 object를 사용한 이유가 궁금해요~

Kimprodp replied:
@skylar1220 object는 override 메서드라 못바꾸지 않나요?

hyeonji1220 replied:
앗 설명이 부족했네요
`RedisTemplate<String, Object> redisTemplate;` 부분을 말합니다!

nayonsoso replied:
레디스의 '사용 확장성'을 생각했어요!
레디스에 반드시 Integer 이나 Long 과 같은 값을 담을게 아니라,
"세션의 정보"와 같은 것들도 담을 수 있게 될 것 같아요.
(무중단 배포로 sticky session 을 사용하지 않게 되면 곧 우리가 해야 할 일이기도 하고요😅)

그런데 지금 숫자로 value 를 한정하면, 나중에 다른 형태의 값을 저장할 때 기존 기능에 영향을 주게 돼요.
그래서 Object 를 저장하게 해준 것입니다!

nayonsoso replied:
추가))

저는 어플리케이션에 하나의 RedisTemplate 을 추가할 수 있을거라 생각했는데, 알고보니 아니었네요🫨 세션같은 정보를 저장할때는 그에 맞는 형식의 RedisTemplate 을 추가하도록 하고, 중복 요청정보를 저장할 때는 `RedisTemplate<String, Long>` 로 저장하도록 하겠습니다!

그리고 여러개의 RedisTemplate을 사용하더라도 저장을 할 때는 하나의 레디스에 key 와 value 모두 String 으로 직렬화해서 사용해서 상관 없다고 합니다😊 커비가 짚어준 덕분에 더 효율적으로 코드를 작성하게 되었어요!!!! 커비 최고최고😍

hyeonji1220 commented on backend/src/test/java/reviewme/global/DuplicateRequestInterceptorTest.java:
```diff
@@ -0,0 +1,99 @@
+    @Test
+    void 특정_POST_요청이_처음이_아니며_최대_빈도보다_클_경우_예외를_발생시킨다() {
+        // given
+        int maxFrequency = 3;
```
3부터 예외를 반환하니 maxFrequency를 경계값인 2로 하는 건 어떨까요?

nayonsoso replied:
3번째 요청까지는 받고, 4번째 요청부터 리젝합니다!
그래서 `maxFrequency = 3`인 현채 코드가 경계값이라 생각해요

hyeonji1220 reviewed (changes requested):
산따최!!
어려워보이는데 구현해냈고나🚀🚀

donghoony commented on backend/src/main/java/reviewme/global/DuplicateRequestInterceptor.java:
```diff
@@ -0,0 +1,55 @@
+
+@Component
+@RequiredArgsConstructor
+public class DuplicateRequestInterceptor implements HandlerInterceptor {
```
인터셉터 차원에서 로깅할 필요는 없을까요? 한계치를 넘어가면 WARN 정도로 로그 찍어주면 좋겠어요!

nayonsoso replied:
> 인터셉터 차원에서 로깅할 필요는 없을까요?

예외에서 `Too many duplicate request received` 로 로그를 찍고 있습니다👍 이것 말고도 추가적인 로그가 필요하다 생각하시나요?

> 한계치를 넘어가면 WARN 정도로 로그 찍어주면 좋겠어요!

로그 레벨을 info -> warn 으로 바꿔주었습니다😊

donghoony replied:
충분하겠습니다~

donghoony commented on backend/src/main/java/reviewme/global/DuplicateRequestInterceptor.java:
```diff
@@ -0,0 +1,55 @@
+    private String generateRequestKey(HttpServletRequest request) {
+        String requestURI = request.getRequestURI();
+        String remoteAddr = request.getRemoteAddr();
+        String userAgent = request.getHeader("User-Agent");
```
`HttpHeaders.USER_AGENT` 상수가 있습니다 👍🏻

nayonsoso replied:
음~ 맛있다~😋

donghoony commented on backend/src/main/java/reviewme/global/DuplicateRequestInterceptor.java:
```diff
@@ -0,0 +1,55 @@
+            return true;
+        }
+
+        if (!(value instanceof Integer)) {
+            throw new RequestFrequencyNonNumericException(value);
+        }
```
코드상으로 절대 일어날 수 없는 사항같은데, 무시해도 되지 않을까요? 클라이언트의 악의적인 요청 맥락이 아니라서, 충분히 저희가 핸들링할 수 있다고 생각해요.

Kimprodp commented on backend/src/main/java/reviewme/config/WebConfig.java:
```diff
@@ -1,16 +1,28 @@
+
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        registry.addInterceptor(new DuplicateRequestInterceptor(redisTemplate));
```
요긴 경로 설정 따로 안하면 디폴트가 전체 경로인가요?

nayonsoso replied:
네! 디폴트가 모든 경로로 알고 있어요.
그래도 명시하는편이 좋을까요?

Kimprodp replied:
아뇨 몰랐어용 지금 괜찮다고 봅니다!

Kimprodp commented on backend/src/main/java/reviewme/global/exception/TooManyDuplicateRequestException.java:
```diff
@@ -0,0 +1,12 @@
+public class TooManyDuplicateRequestException extends ReviewMeException {
+
+    public TooManyDuplicateRequestException(String requestKey) {
+        super("짧은 시간 안에 너무 많은 동일한 요청이 일어났어요");
```
요청 제한 로직에 대한 정보를 클라이언트에 응답해도 괜찮을까요?  유추할 수 있는 정보는 제외하면 좋을 것 같다는 생각입니다~

nayonsoso replied:
저는 이 메세지가 "요청 제한 로직에 대한 정보"라기보다,
"너무 많이 요청했으니 잠시 있다가 다시 시도해주세요" 라는 식의 메세지라고 봤어요!
사용자를 도와주는 메세지라 괜찮다고 봅니다👀

Kimprodp replied:
그렇다면요 슨초 말대로 정상적인 요청이지만 어떤 이유로 중복되어서 짧은 시간 안에 요청이 들어온 경우 다시 안내하는 것을 의도하는 것이니, 마지막에 "잠시 후 다시 시도해주세요" 라는 문구를 추가하면 더 좋을 것 같아요!

nayonsoso replied:
반완입니다~

Kimprodp commented on backend/src/main/java/reviewme/global/DuplicateRequestInterceptor.java:
```diff
@@ -0,0 +1,55 @@
+        }
+        int frequency = (int) value;
+        if (frequency >= MAX_FREQUENCY) {
+            throw new TooManyDuplicateRequestException(key);
```
맥락상 key 말고 frequency 를 전달하는 것이 더 읽는 데 자연스러운 것 같아요~

nayonsoso replied:
key 를 전달하면 이렇게 로깅되는데,

Too many duplicate request received: RequestURI: /v2/groups, RemoteAddr: 127.0.0.1, UserAgent: Apache-HttpClient/4.5.14 (Java/21.0.4)

value(frequency) 를 전달하면 이렇게 로깅됩니다.

Too many duplicate request received: 4

key 를 전달하는게 더 맥락을 잘 전달한다고 생각해요~

Kimprodp reviewed (changes requested):
산초 커밋과 테스트를 읽으니 깔끔하게 이해가 잘 되네요👍👍
고생했습니다 코멘트만 확인 부탁해요~

donghoony commented on backend/src/main/java/reviewme/global/DuplicateRequestInterceptor.java:
```diff
@@ -0,0 +1,55 @@
+            return true;
+        }
+
+        String key = generateRequestKey(request);
+        Object value = redisTemplate.opsForValue().get(key);
+        if (value == null) {
+            redisTemplate.opsForValue().set(key, 1, DURATION);
+            return true;
+        }
```
중간에 분기를 일으키지 않고, 아래와 같이 하면 어떤가요?
```java
redisTemplate.opsForValue().setIfAbsent(key, 0, DURATION);
redisTemplate.opsForValue().increment(key);
requestCount = redisTemplate.opsForValue().get(key);
```

nayonsoso replied:
음~~ setIfAbsent 맛있다😋
제공되는 함수들로 로직을 개선할 수 있는지 더 꼼꼼히 확인해야겠네요😅
반영했습니다!

![image](https://github.com/user-attachments/assets/f4d10a97-0057-4bb5-8858-d48f6f3c3c09)
이런 노란줄이 뜨긴 하는데, 로직상 발생할 수 없으니 괜찮겠죠?

donghoony replied:
바로 위에서 `setIfAbsent`를 해서 괜찮다고 생각하기는 해요 🤔

nayonsoso committed:
style: 파일 끝 개행

nayonsoso committed:
refactor: ConfigurationProperties 적용

nayonsoso committed:
refactor: 변수명 변경

nayonsoso committed:
refactor: 기존 상수 적용

nayonsoso committed:
refactor: RedisTemplate 타입과 중복 요청 감지 로직 변경

- RedisTemplate<String, Object> -> RedisTemplate<String, Long>

nayonsoso committed:
refactor: 로그 메세지 수정

nayonsoso committed:
chore: 사용하지 않는 예외 삭제

nayonsoso committed:
refactor: 로그 레벨 수정

- info -> warn

donghoony commented on backend/src/main/java/reviewme/global/DuplicateRequestInterceptor.java:
```diff
@@ -0,0 +1,55 @@
+public class DuplicateRequestInterceptor implements HandlerInterceptor {
+
+    private static final int MAX_FREQUENCY = 3;
+    private static final Duration DURATION = Duration.of(1, ChronoUnit.SECONDS);
```
MAX_FREQ, Duration도 config로 옮기면 어떤가요?

nayonsoso replied:
반완완~~

nayonsoso committed:
refactor: 중복 요청 검증 로직 개션

- if null 분기 제거

nayonsoso committed:
test: 깨진 테스트 봉합

nayonsoso committed:
test: 의미 없어진 테스트 삭제

- 명시적으로 null 인 경우 1로 초기화했던 이전 코드와 달리, setIfAbsent를 통해서 값이 초기화하는 지금은 '1로 초기화되었는지' 검증하기가 어렵다.

nayonsoso committed:
refactor: 불필요한 지역변수 할당 삭제

- increment 가 증가한 결과를 바로 반환한다. 이를 사용하도록 수정했다.

nayonsoso committed:
test: redisTemplate 등록으로 깨지는 테스트 봉합

- RedisTemplate과 ValueOperations를 모킹한다.
- 구체적인 인자를 지정하지 않은 stub 문을 수정한다.

nayonsoso commented on backend/src/test/resources/application.yml:
```diff
@@ -38,3 +38,8 @@ logging:
 cors:
   allowed-origins:
     - https://allowed-domain.com
+
+# todo: 이게 맞는걸까..?
+redis:
+  host: localhost
+  port: 6379
```
이 설정은 test/application.yml 에 들어갑니다.
딱히 의미 있는 설정은 아니지만, 
스프링 부트 테스트를 할 때 "빈등록" 성공을 위해서 작성해주었습니다.
(어플리케이션에서 yml 로부터 값을 읽어와서 빈을 등록하니깐요~)
조금 찝찝하긴 한데 이게 최선이라 생각해서 이렇게 해놨어요😅

donghoony replied:
(반영하라는 건 아니예요~) 테스트용 configuration에 `Properties`를 새로 만든 `Template`을 등록할 수도 있습니다

nayonsoso committed:
refactor: 예외 이름 변경

Kimprodp reviewed (approved):
신초 고생했읍니다. 빨리 적용하고 중복요청 보내보고 싶어요 👍👍

nayonsoso committed:
refactor: 요청 제한 설정을 properties로 이동

donghoony reviewed (approved):
따닥 미워

hyeonji1220 reviewed (approved):
어렵다.. 산초 최고🚀🚀🚀🚀

nayonsoso committed:
refactor: 예외 메세지 수정

- 테드 의견 반영

nayonsoso committed:
chore: 안쓰는 import 문 제거

nayonsoso committed:
test: 테스트 데이터 경계값으로 변경

nayonsoso committed:
refactor: 제한 만료 시간 설정 로직 변경

nayonsoso committed:
refactor: frequency -> requestCount 완전 대체
